templates/rustdoc.hbs

```diff
@@ -9,26 +9,35 @@
     <link rel="search" href="/opensearch.xml" type="application/opensearchdescription+xml" title="Docs.rs">
 </head>
 <body>
-{{> navigation_rustdoc}}
-<div id="rustdoc_body_wrapper" class="{{content.rustdoc_body_class}}" tabindex="-1">
-    {{{content.rustdoc_body}}}
-</div>
-<script type="text/javascript" src="/menu.js?{{cratesfyi_version_safe}}"></script>
+    <div id="page-container">
+        {{> navigation_rustdoc}}
+        <div id="rustdoc_body_wrapper" class="{{content.rustdoc_body_class}}" tabindex="-1">
+            {{{content.rustdoc_body}}}
+        </div>
+    </div>
+    <script type="text/javascript" src="/menu.js?{{cratesfyi_version_safe}}"></script>
 </body>
 <script>
-    var doc_body = document.getElementById("rustdoc_body_wrapper");
-    if(window.location.hash) {
-        var notFirefox = typeof InstallTrigger === 'undefined';
-        if(notFirefox) {
-            var hash = window.location.hash;
-            window.location.hash = "";
-            setTimeout(function () {
-                window.location.hash = hash;
-                doc_body.focus();
-            }, 1);
-        }
-    } else {
-        doc_body.focus();
-    }
+ // restore scroll state across back/forward navigation & give focus
+ // to the scrolling container:
+ (function() {
+     var keyName = function() {
+         return 'rustdoc_body_scroll_posn_' + window.location.href;
+     }
+     var oldPosn = window.sessionStorage.getItem(keyName());
+     var wrapper = document.getElementById('rustdoc_body_wrapper');
+     var saveState = function() {
+         window.sessionStorage.setItem(keyName(), wrapper.scrollTop);
+     }
+     if(oldPosn) {
+         wrapper.scrollTop = oldPosn;
+     } else {
+         // we have no scroll state saved, set it now:
+         saveState();
+     }
+     wrapper.addEventListener("scroll", saveState);
+     window.addEventListener("hashchange", saveState);
+     wrapper.focus();
+ })()
```

Comment:
This is in an IIFE so that the variables aren't added to the global scope?

Reply:
Yup, I was getting worried that the previous change I'd had in mind used too many vars.